### PR TITLE
ENH/REF: refactor shared table views

### PR DESCRIPTION
**Pull request by tangkong**
## Description
- Refactors shared table models to have their own self-contained table views for ease of use and extensibility
- Add methods to make columns editable for both shared table models
- Adds ValueDelegate to provide constrained editing for columns based on their datatype and metadata
- Add more metadata fields to EpicsData (precision, enums, limits, precision)
- improve enum handling for `LivePVTableView`

## Motivation and Context
Going about adding more functionality.

## How Has This Been Tested?
Interactively, ensuring existing tests run

## Where Has This Been Documented?
This PR
enum ValueDelegate
image

double ValueDelegate (must be < 0)
image

## Pre-merge checklist

- [x] Code works interactively
- [x] Code follows the [style guide](https://pcdshub.github.io/style.html)
- [x] Code contains descriptive docstrings, including context and API
- [x] New/changed functions and methods are covered in the test suite where possible
- [x] Test suite passes locally
- [x] Test suite passes on GitHub Actions
- [x] Ran ``docs/pre-release-notes.sh`` and created a pre-release documentation page
- [x] Pre-release docs include context, functional descriptions, and contributors as appropriate

### CODE REVIEW

**Comment by tangkong** [comment]
Not requesting a review from Devan, as he's on vacation.  This was starting to snowball (as always), so I'll cut this off here.  Everything was originally in service of making the cells editable.  Lots of little things got hit along the way

**Review by ZLLentz — COMMENTED** [review]
Partial review: I'll need to start again at the views definitions when I feel less tired

**Comment by ZLLentz** [comment]
The result looks great though, next time I'll try to run it myself and read through the code with renewed energy

**Comment by ZLLentz** [comment]
@tangkong what's the best way for me to sun superscore and try this out myself?

**Comment by tangkong** [comment]
This should work
```bash
$ export SUPERSCORE_CFG=/path/to/repo/superscore/superscore/tests/config.cfg
$ python -m superscore ui
```

File: superscore/widgets/views.py [hunk 1]
```diff
@@ -529,16 +566,30 @@ def flags(self, index: QtCore.QModelIndex) -> QtCore.Qt.ItemFlag:
         QtCore.Qt.ItemFlag
             the ItemFlag corresponding to the cell
         """
-        if (index.column() == len(self.headers) - 1):
-            return QtCore.Qt.ItemIsEditable | QtCore.Qt.ItemIsEnabled
+        if self._editable_cols[index.column()]:
```
Comment (ZLLentz) [inline superscore/widgets/views.py:569]: My GUI crashes here if I open `python -m superscore ui` and then click on `Apply Filter` with no other inputs
```
$ python -m superscore ui
qt.glx: qglx_findConfig: Failed to finding matching FBConfig for QSurfaceFormat(version 2.0, options QFlags<QSurfaceFormat::FormatOption>(), depthBufferSize -1, redBufferSize 1, greenBufferSize 1, blueBufferSize 1, alphaBufferSize -1, stencilBufferSize -1, samples -1, swapBehavior QSurfaceFormat::SingleBuffer, swapInterval 1, colorSpace QSurfaceFormat::DefaultColorSpace, profile  QSurfaceFormat::NoProfile)
No XVisualInfo for format QSurfaceFormat(version 2.0, options QFlags<QSurfaceFormat::FormatOption>(), depthBufferSize -1, redBufferSize 1, greenBufferSize 1, blueBufferSize 1, alphaBufferSize -1, stencilBufferSize -1, samples -1, swapBehavior QSurfaceFormat::SingleBuffer, swapInterval 1, colorSpace QSurfaceFormat::DefaultColorSpace, profile  QSurfaceFormat::NoProfile)
Falling back to using screens root_visual.
Traceback (most recent call last):
  File "/cds/home/z/zlentz/github/superscore/superscore/widgets/views.py", line 569, in flags
    if self._editable_cols[index.column()]:
       ~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^
KeyError: 0
Aborted (core dumped)
```

Comment (ZLLentz) [reply superscore/widgets/views.py:569]: Actually this might just be py3.12 incompat, nothing seems to work properly, I set up a new env for superscore things and I foolishly picked latest

Comment (tangkong) [reply superscore/widgets/views.py:569]: Well that's unfortunate.  I'll take a looksee

Comment (tangkong) [reply superscore/widgets/views.py:569]: also the page I was testing against was the collection builder page, which you can navigate to via File > New > Collection.  

This also crashes here for me.  I forgot that the search page also uses these table views whoops

**Comment by ZLLentz** [comment]
Nothing in the code stands out as odd and there's a lot of qt stuff that's hard to parse through, so instead I'm going to reset my environment and approach this from a qa perspective

File: superscore/widgets/views.py (hunk 1 above)
Comment (tangkong) [reply superscore/widgets/views.py:572]: Should be fixed in https://github.com/pcdshub/superscore/pull/90/commits/891849f2a9a5c3d9e07a34424137c88220395746

**Comment by ZLLentz** [comment]
I'm not sure what I'm doing wrong, maybe I don't understand how to use the ui. I can't view any of the configs.

**Comment by tangkong** [comment]
We don't have entry-specific pages for anything but the `Collection` yet, which itself is just a placeholder.  Right now the only pages are the collection builder page and the search page.

What are you trying to get to that isn't working?

**Comment by ZLLentz** [comment]
I guess I was expecting to be able to open a page and see one of these views in action

**Comment by tangkong** [comment]
The collection builder page has these views and should be working.  
image

Sorry I haven't been the most clear with how to test this here.  If you navigate File -> New -> Collection, the collection-builder-page will open, from which you can add PVs and sub-collections.  This by default doesn't let you edit the entries added, but the type-specific delegates should appear just find

**Comment by ZLLentz** [comment]
Ok I clicked around on the new collection gui
I couldn't find where to see the "stored value" thing I think because I couldn't figure out how to make a snapshot

Feel free to ignore any of these if they are already known or are pre-existing issues

- Clicking tags+ crashes the gui
```
Traceback (most recent call last):
  File "/cds/home/z/zlentz/github/superscore/superscore/widgets/core.py", line 216, in add_tag
    elem = tags_list.add_item('')
  File "/cds/home/z/zlentz/github/superscore/superscore/widgets/core.py", line 297, in add_item
    self.data_list.append(starting_value)
AttributeError: 'QDataclassValueForobject' object has no attribute 'append'
Aborted (core dumped)
```
- I get lots of qt.glx errors but it might be my environment
```
qt.glx: qglx_findConfig: Failed to finding matching FBConfig for QSurfaceFormat(version 2.0, options QFlags<QSurfaceFormat::FormatOption>(), depthBufferSize -1, redBufferSize 1, greenBufferSize 1, blueBufferSize 1, alphaBufferSize -1, stencilBufferSize -1, samples -1, swapBehavior QSurfaceFormat::SingleBuffer, swapInterval 1, colorSpace QSurfaceFormat::DefaultColorSpace, profile  QSurfaceFormat::NoProfile)
No XVisualInfo for format QSurfaceFormat(version 2.0, options QFlags<QSurfaceFormat::FormatOption>(), depthBufferSize -1, redBufferSize 1, greenBufferSize 1, blueBufferSize 1, alphaBufferSize -1, stencilBufferSize -1, samples -1, swapBehavior QSurfaceFormat::SingleBuffer, swapInterval 1, colorSpace QSurfaceFormat::DefaultColorSpace, profile  QSurfaceFormat::NoProfile)
Falling back to using screens root_visual.
```
- It'd be nice to have a way to edit PVs that have already been added to a collection before saving it, I instead need to delete and add a new PV
- Hitting save configuration twice gives this (the first save didn't seem to add it to the tree or make it findable via search). Maybe save configuration should close the edit gui or replace it with the view-only gui?
```
ERROR:superscore.widgets.views:Failed to set data (False) ->(0, 9): <LivePVHeader.REMOVE: 9>
ERROR:superscore.widgets.views:Failed to set data (False) ->(0, 8): <LivePVHeader.OPEN: 8>
INFO:superscore.widgets.page.collection_builder:Collection saved (10c5933d-79a3-46ff-8bd3-96a1352e3bf1)
qt.glx: qglx_findConfig: Failed to finding matching FBConfig for QSurfaceFormat(version 2.0, options QFlags<QSurfaceFormat::FormatOption>(), depthBufferSize -1, redBufferSize 1, greenBufferSize 1, blueBufferSize 1, alphaBufferSize -1, stencilBufferSize -1, samples -1, swapBehavior QSurfaceFormat::SingleBuffer, swapInterval 1, colorSpace QSurfaceFormat::DefaultColorSpace, profile  QSurfaceFormat::NoProfile)
No XVisualInfo for format QSurfaceFormat(version 2.0, options QFlags<QSurfaceFormat::FormatOption>(), depthBufferSize -1, redBufferSize 1, greenBufferSize 1, blueBufferSize 1, alphaBufferSize -1, stencilBufferSize -1, samples -1, swapBehavior QSurfaceFormat::SingleBuffer, swapInterval 1, colorSpace QSurfaceFormat::DefaultColorSpace, profile  QSurfaceFormat::NoProfile)
Falling back to using screens root_visual.
Traceback (most recent call last):
  File "/cds/home/z/zlentz/github/superscore/superscore/widgets/page/collection_builder.py", line 124, in save_collection
    self.client.save(self.data)
  File "/cds/home/z/zlentz/github/superscore/superscore/client.py", line 174, in save
    self.backend.save_entry(entry)
  File "/cds/home/z/zlentz/github/superscore/superscore/backends/filestore.py", line 271, in save_entry
    raise BackendError("Entry already exists, try updating the entry "
superscore.errors.BackendError: Entry already exists, try updating the entry instead of saving it
Aborted (core dumped)
```
- After saving my config, closing, and re-opening, I now have a validation error on opening the config again
```
$ python -m superscore ui
Traceback (most recent call last):
  File "/cds/home/z/zlentz/miniconda3/envs/superscore/lib/python3.9/runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/cds/home/z/zlentz/miniconda3/envs/superscore/lib/python3.9/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/cds/home/z/zlentz/github/superscore/superscore/__main__.py", line 3, in <module>
    main()
  File "/cds/home/z/zlentz/github/superscore/superscore/bin/main.py", line 97, in main
    func(**kwargs)
  File "/cds/home/z/zlentz/github/superscore/superscore/bin/ui.py", line 21, in main
    main_window = Window()
  File "/cds/home/z/zlentz/github/superscore/superscore/widgets/window.py", line 45, in __init__
    self.setup_ui()
  File "/cds/home/z/zlentz/github/superscore/superscore/widgets/window.py", line 56, in setup_ui
    self.tree_model = RootTree(base_entry=self.client.backend.root,
  File "/cds/home/z/zlentz/github/superscore/superscore/backends/filestore.py", line 256, in root
    with self._load_and_store_context():
  File "/cds/home/z/zlentz/miniconda3/envs/superscore/lib/python3.9/contextlib.py", line 119, in __enter__
    return next(self.gen)
  File "/cds/home/z/zlentz/github/superscore/superscore/backends/filestore.py", line 353, in _load_and_store_context
    db = self._load_or_initialize()
  File "/cds/home/z/zlentz/github/superscore/superscore/backends/filestore.py", line 63, in _load_or_initialize
    self._root = self.load()
  File "/cds/home/z/zlentz/github/superscore/superscore/backends/filestore.py", line 251, in load
    return deserialize(Root, serialized)
  File "/cds/home/z/zlentz/miniconda3/envs/superscore/lib/python3.9/site-packages/apischema/deserialization/__init__.py", line 887, in deserialize
    return deserialization_method(
  File "/cds/home/z/zlentz/miniconda3/envs/superscore/lib/python3.9/site-packages/apischema/deserialization/methods.py", line 697, in deserialize
    raise ValidationError(errors or [], field_errors or {})
apischema.validation.errors.ValidationError: ValidationError: [{'loc': ['entries', 4, 'Collection', 'description'], 'err': 'expected type string, found array'}, {'loc': ['entries', 4, 'Collection', 'title'], 'err': 'expected type string, found array'}]
```
Here's the diff from my db:
```
+    },
+    {
+      "Collection": {
+        "uuid": "10c5933d-79a3-46ff-8bd3-96a1352e3bf1",
+        "description": [
+          "Defo neater than your collection haha"
+        ],
+        "creation_time": "2024-10-09T23:28:21.315755+00:00",
+        "title": [
+          "My neat collection"
+        ],
+        "children": [
+          {
+            "uuid": "3de82331-0f13-4ed7-a668-0d33eae8cd1e",
+            "description": "",
+            "creation_time": "2024-10-09T23:32:00.149681+00:00",
+            "pv_name": "IM3L0:PPM:MMS:STATE:SET",
+            "abs_tolerance": null,
+            "rel_tolerance": null,
+            "readback": {
+              "uuid": "c092bad8-ada1-4efa-abd8-46306b0b7836",
+              "description": "",
+              "creation_time": "2024-10-09T23:32:00.149661+00:00",
+              "pv_name": "IM3L0:PPM:MMS:STATE:GET_RBV",
+              "abs_tolerance": null,
+              "rel_tolerance": null,
+              "readback": null,
+              "read_only": true
+            },
+            "read_only": false
+          }
+        ],
+        "tags": []
+      }
```
Note how the description and title are arrays of strings instead of strings

Anyway sorry for only having bugs to report, I think the snapshot gui from your screenshot above looks the most snazzy but I couldn't figure out how to get to it

**Comment by tangkong** [comment]
No, thank you for the bug reports.  Helps keep me honest.  The editable delegates aren't immediately accessible, since the existing pages don't support editing values by design.  I'm currently working on some additional pages that should expose this more clearly (hopefully)

I'll go through the bugs and see what I can fix. 👍

**Comment by tangkong** [comment]
There were stray commas somehow, making strings into tuples.  Validation errors and saving title/descriptions into lists should be fixed in https://github.com/pcdshub/superscore/pull/90/commits/ef4bc1f5a53dc43bad280ba1d67de9cd658a4f3c

**Comment by tangkong** [comment]
Tags will remain broken until we actually implement the feature.  Part of the problem is that the `TagsWidget` was expecting the dataclass field to be a list not a set, and the other part of the problem is that the field is a `Set[Tag]`, where `Tag` is a `Flag`.  We haven't tested serialization with these flags, and if we intend to centrally manage / restrict the tags that can be chosen, this widget will need further modifications.  

All this to say I'm not going to work on this further for now.  😢

**Comment by tangkong** [comment]
If you're interested in running a slightly  modified collection builder page to see the delegates work, you can run this file:
```
python /cds/home/r/roberttk/test/superscore/enh_views_test.py
```

**Comment by tangkong** [comment]
It has become abundantly clear that this was written during the heat wave.  The quality is awful.  Thank you for being patient.  I've got many test cases to write, now that I've caught so many little bugs

**Comment by ZLLentz** [comment]
I think this is probably mature enough to merge with some documentation of things to be fixed later

**Comment by tangkong** [comment]
I'll make some issues and add some tests.

File: superscore/tests/test_views.py [hunk 2]
```diff
@@ -32,6 +37,42 @@ def pv_poll_model(
...
+    LivePVTableView, holds three PVs with different types.  Stored data allowed.
+    Mocks control layer to return realistic-ish EpicsData
+
+    TODO: add string field examples once we properly handle strings
+    """
+    # Build side effect function:
+    ret_vals = {
+        "MY:FLOAT": EpicsData(data=0.5, precision=3,
+                              upper_ctrl_limit=2, lower_ctrl_limit=-2),
+        "MY:INT": EpicsData(data=1, upper_ctrl_limit=10, lower_ctrl_limit=-10),
```
Comment (ZLLentz) [inline superscore/tests/test_views.py:56]: The most nitpick I've ever nitpicked: high, low ranges break my brain and I'd prefer low, high
```suggestion
                              lower_ctrl_limit=-2, upper_ctrl_limit=2),
        "MY:INT": EpicsData(data=1, lower_ctrl_limit=-10, upper_ctrl_limit=10),
```

File: superscore/tests/test_views.py [hunk 3]
```diff
@@ -32,6 +37,42 @@ def pv_poll_model(
...
+
+    view = LivePVTableView()
+    view.client = mock_client
+    view.set_data(simple_snapshot)
+
+    qtbot.wait_until(lambda: view.model()._poll_thread.isRunning())
+    yield view
+
+    view.model().stop_polling()
+    qtbot.wait_until(lambda: not view.model()._poll_thread.isRunning())
```
Comment (ZLLentz) [inline superscore/tests/test_views.py:73]: This test has no assert statements. Is there nothing specific to test for in the basic behavior of getting values?

**Review by ZLLentz — APPROVED** [review]
I think I only had testing nitpicks

**Review by ZLLentz — APPROVED** [review]

**Review by shilorigins — APPROVED** [review]

**Comment by tangkong** [comment]
Thanks everyone for the reviews!
